interface/src/app/gene/genes.js

```diff
@@ -0,0 +1,377 @@
+  return {
+    controller: ['$scope', '$rootScope', 'SearchResults',
+      function($scope, $rootScope, SearchResults) {
+
+        $scope.loadingSearchResults = false;
+
+        // Clear any existing search results
+        SearchResults.clear();
+
+        $scope.searchGenes = function() {
+          if (!$scope.genesToAdd) { //if the query is empty
+            return false;
+          }
+          $scope.loadingSearchResults = true;
+
+          $rootScope.$broadcast( 'results.loadingSearchResults' );
+          qparams = {'query': $scope.genesToAdd.query};
+
+          if ($scope.organism) {
+            qparams['organism'] = $scope.organism;
+          }
+
+          SearchResults.search(qparams);
+        };
+
+        $scope.$on('results.searchResultsReturned', function() {
+          $scope.loadingSearchResults = false;
+        });
+
+      }
+    ],
+    replace: true,
+    restrict: "E",
+    scope: {
+        query: '@'
+    },
+    templateUrl: 'gene/gene-search-form.tpl.html'
+  };
+}])
+
+// Directive for table containing search results
+.directive('searchResultTable', ['SearchResults', function( SearchResults ) {
+  return {
+    controller: ['$scope', 'SearchResults', function($scope, SearchResults) {
+      $scope.currentPage = 1;
+      $scope.itemsPerPage = 10;
+      $scope.totalResults = 0;
+      $scope.maxSize = 10;
+      $scope.resultsForPage = [];
+      $scope.searchResults = SearchResults.getSearchResults();
+      $scope.loadingSearchResults = false;
+
+      $scope.addAllNonAmbiguous = function() {
+        // Function to automatically add all genes that
+        // only have one search result.
+        var searchResults = SearchResults.getSearchResults();
+        for (var key in searchResults) {
+          var results = searchResults[key];
+          if (results['found'].length <= 1) {
+            if (results['found'][0]) {
+              SearchResults.remove(key);
+            }
+          }
+        }
+      };
+
+      $scope.removeNotFound = function() {
+          // Function to get rid of all the queries that returned no results.
+          var searchResults = SearchResults.getSearchResults();
+          for (var key in searchResults) {
+            var results = searchResults[key];
+            if (results['found'].length === 0) {
+              SearchResults.remove(key);
+            }
+          }
+      };
+    }],
+
+    link: function(scope, element, attr) {
+      scope.$on('results.update', function() {
+        scope.totalResults = SearchResults.size();
+        var begin = ((scope.currentPage - 1) * scope.itemsPerPage);
+        var end = begin + scope.itemsPerPage;
+        scope.resultsForPage = SearchResults.getQueries().slice(begin, end);
+      });
+
+      scope.$on('results.loadingSearchResults', function() {
+        scope.loadingSearchResults = true;
+      });
+
+      scope.$on('results.searchResultsReturned', function() {
+        scope.loadingSearchResults = false;
+      });
```

Comment:
Okay, I see now where these signals are getting picked up. Maybe I'm missing something, but it feels non-Angularish to be using a `$broadcast` to set a variable like this. The `loadingSearchResults` variable should exist in one place as the primary model and any view that needs to access it should refer to that instance and let two-way binding handle the rest.

Another possibility is to use a chain of `$promise`s to achieve the sequence of updates required.